### PR TITLE
Project: Knight's Travails

### PR DESCRIPTION
**Pull request by Kaktug**
# Description

This PR contains my solution to The Odin Project's "Knight's Travails" project.  The details of the assignment can be viewed here:
https://www.theodinproject.com/lessons/ruby-knights-travails

> ## Assignment
> A knight in chess can move to any square on the standard 8x8 chess board from any other square on the board, given enough turns. Its basic move is two steps forward and one step to the side. It can face any direction.
>
> Your task is to build a function `#knight_moves` that shows the shortest possible way to get from one square to another by outputting all squares the knight will stop on along the way.
>
> You can think of the board as having 2-dimensional coordinates. Your function would therefore look like:
> - `knight_moves([0,0],[1,2]) == [[0,0],[1,2]]`
> - `knight_moves([0,0],[3,3]) == [[0,0],[1,2],[3,3]]`
> - `knight_moves([3,3],[0,0]) == [[3,3],[1,2],[0,0]]`
> 1. Put together a script that creates a game board and a knight.
> 1. Treat all possible moves the knight could make as children in a tree. Don’t allow any moves to go off the board.
> 1. Decide which search algorithm is best to use for this case. Hint: one of them could be a potentially infinite series.
> 1. Use the chosen search algorithm to find the shortest path between the starting square (or node) and the ending square. Output what that full path looks like, e.g.:
> ```
> > knight_moves([3,3],[4,3])
>   => You made it in 3 moves!  Here's your path:
>     [3,3]
>     [4,5]
>     [2,4]
>     [4,3]
> ```

# Notes
Like #22 I went ahead and included some test cases in addition to The Odin Project's request for a "driver" script.  I went down a lot of wrong paths with this one.  Figuring out the spiderwork of moves was too much for me to keep in my head, and too hard to follow even when I wrote simple cases out on paper.  I ended up iterating into the answer one-step-at-a-time.  Getting the `possible_next_moves` method was crucial.  Then I figured out the idea of a dual queue (one to track current working state and one to track where the Knight had visited).  I struggled for a while to figure out where ownership of the queues should go.  I started with the Knight but ended up moving that responsibility to the board because it made things wayyy easier (where a "node" is essentially equivalent to a "square" on the board).  That was the really hard part.

Then from there I reasoned we could find the node associated with the destination coordinates and traverse up its parent tree to figure out the move path.  That part was pretty easy, and it was nice to end a real head-scratcher of a project with a straightforward denouement.

### CODE REVIEW

**Review by chmurph2 — APPROVED** [review]
Stared at all of this for a while and it LGTM. I left some suggestions for you to consider.

File: full-stack-ruby-on-rails/project-knights-travails/lib/knight.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    @coordinates = starting_position
+    @board_size = board_size - 1
+  end
+
+  def possible_next_moves(visited_squares)
+    x, y = @coordinates
+    MOVES.map { |x_offset, y_offset| [x + x_offset, y + y_offset] }
+         .reject { |move| out_of_bounds?(move) }
+         .reject { |move| already_visited?(move, visited_squares) }
+  end
```
[the remark this answers is not in the document]
Comment (chmurph2) [reply full-stack-ruby-on-rails/project-knights-travails/lib/knight.rb:20]: Maybe a `Move` is a kind of `Node`?

File: full-stack-ruby-on-rails/project-knights-travails/spec/game_board_spec.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+      k = b.knight_moves([0, 0], [1, 2])
+      expected_path = [[0, 0], [1, 2]]
+      expected_move_count = expected_path.size - 1
+
+      expect(k[:move_count]).to eq expected_move_count
+      expect(k[:move_count] + 1).to eq k[:pathway].length
+      expect(k[:pathway]).to eq expected_path
+
+      k = b.knight_moves([0, 0], [3, 3])
```
[the remark this answers is not in the document]
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-knights-travails/spec/game_board_spec.rb:18]: Good timing as I just got started with some of the RSpec exercises.  Those suggestions sound good to me, I'll adjust!

File: full-stack-ruby-on-rails/project-knights-travails/lib/knight.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kaktug) [reply full-stack-ruby-on-rails/project-knights-travails/lib/knight.rb:20]: It's not immediately obvious to me how I would implement it but I will spend some time thinking about it.  I'm sure this will probably come up in some capacity for the upcoming Chess assignment 😅

[the remark this answers is not in the document]
Comment (chmurph2) [reply full-stack-ruby-on-rails/project-knights-travails/lib/knight.rb:20]: We can pair on it too, whenever.
